book/src/rfcs/foreign_field_mul.md

```diff
@@ -0,0 +1,1273 @@
+
+So, the equation we need to constrain is
+
+$$
+x + f' = x'.
+$$
+
+We can expand the left hand side into the 2 limb format in order to obtain 2 intermediate sums
+
+$$
+\begin{aligned}
+s_{01} = x_{01} + f_{01}' \\
+s_2 = x_2 + f'_2 \\
+\end{aligned}
+$$
+
+where $x_{01}$ and $f'_{01}$ are defined like this
+
+$$
+\begin{aligned}
+x_{01} = x_0 + 2^{\ell} \cdot x_1 \\
+f'_{01} = f'_0 + 2^{\ell} \cdot f'_1 \\
+\end{aligned}
+$$
+
+and $x$ and $f'$ are defined like this
+
+$$
+\begin{aligned}
+x = x_{01} + 2^{2\ell} \cdot x_2 \\
+f' = f'_{01} + 2^{2\ell} \cdot f'_2 \\
+\end{aligned}
+$$
+
+Going back to our intermediate sums, the maximum bit length of sum $s_{01}$ is computed from the maximum bit lengths of $x_{01}$ and $f'_{01}$
+
+$$
+\underbrace{(2^{\ell} - 1) + 2^{\ell} \cdot (2^{\ell} - 1)}_{x_{01}} + \underbrace{(2^{\ell} - 1) + 2^{\ell} \cdot (2^{\ell} - 1)}_{f'_{01}} = 2^{2\ell+ 1} - 2,
+$$
+
+which means $s_{01}$ is at most $2\ell + 1$ bits long.
+
+Similarly, since $x_2$ and $f'_2$ are less than $2^{\ell}$, the max value of $s_2$ is
+
+$$
+(2^{\ell} - 1) + (2^{\ell} - 1) = 2^{\ell + 1} - 2,
+$$
+
+which means $s_2$ is at most $\ell + 1$ bits long.
+
+Thus, we must constrain
+
+$$
+s_{01} + 2^{2\ell} \cdot s_2 - x'_{01} - 2^{2\ell} \cdot x'_2 = 0 \mod 2^t.
+$$
+
+The accumulation of this into parts looks like this.
+
+```text
+0             L             2L            3L=t          4L
+|-------------|-------------|-------------|-------------|-------------|
+                            :
+|------------s01------------:-| 2L + 1
+                            : ↖w01
+                            |------s2-----:-| L + 1
+                            :               ↖w2
+                            :
+|------------x'01-----------|
+                            :
+                            |------x'2----|
+                            :
+\____________________________/
+             ≈ z01           \_____________/
+                                   ≈ z2
+```
+
+The two parts are computed with
+
+$$
+\begin{aligned}
+z_{01} &= s_{01} - x'_{01} \\
+z_2 &= s_2 - x'_2.
+\end{aligned}
+$$
+
+Therefore, there are two carry bits $w_{01}$ and $w_2$ such that
+
+$$
+\begin{aligned}
+z_{01} &= 2^{2\ell} \cdot w_{01} \\
+z_2 + w_{01} &= 2^{\ell} \cdot w_2
+\end{aligned}
+$$
+
+In this scheme $x'_{01}, x'_2, w_{01}$ and $w_2$ are witness data, whereas $s_{01}$ and $s_2$ are formed from a constrained computation of witness data $x_{01}, x_2$ and constraint system public parameter $f'$.  Note that due to carrying, witness $x'_{01}$ and $x'_2$ can be different than the values $s_{01}$ and $s_2$ computed from the limbs.
```

Comment:
also here, github does not parse it perfectly

Reply:
Yeah :-(. Hopefully they will fix that.  Anyway our plan was to view this in mdbook.